aws/src/main/java/org/apache/iceberg/aws/s3/S3URI.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.aws.s3;
+
+import java.net.URI;
+
+public class S3URI {
+  private final URI uri;
+
+  public S3URI(URI uri) {
+    this.uri = uri;
+  }
+
+  public S3URI(String uri) {
+    this(URI.create(uri));
+  }
+
+  public String bucket() {
+    return uri.getAuthority();
+  }
+
+  public String key() {
+    return uri.getPath().startsWith("/") ? uri.getPath().replaceFirst("/", "") : uri.getPath();
```

Comment:
Unfortunately, Java's `URI` has a lot of problems with character escaping. and it's really hard to reconstruct some or all of a URI from the getter methods that it exposes. The `uriToString` method that we use actually avoids calling `URI#toString` and instead uses `new Path(uri).toString()`.

I think it would be better to parse the URI without it. Since we expect a bucket in the URI authority, we can be fairly strict with parsing rules:
* Split by `://` to get scheme and bucket/key
* Split bucket/key by the first `/` to get bucket and key
* Validate scheme is s3 or equivalent (s3n, s3a)

Reply:
I'm changed this to a custom S3URI implementation that does not rely on URI.  It does basic validation for scheme with know supported types (https, s3, s3a, s3n).  It will also strip query and fragment (some patterns like externally signed urls or other services could leak those into the location).  Beyond that it does no actual validation since the S3 client will do that.